byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs

```diff
@@ -29,6 +34,44 @@ moneyIsConstant = withTests 200 . property $ do
   (st0, st) <- firstAndLastState <$> forAll (trace @UTXOW 500)
   reserves st0 + balance (utxo st0) === reserves st + balance (utxo st)
 
+-- | Check that a UTxO has no common inputs with a transaction
+noCommonInputs :: UTxO -> Tx -> Bool
+noCommonInputs utxo0 tx = dom (txouts tx) ∩ dom utxo0 == empty
+
+-- | Check that there is no double spending
+noDoubleSpending :: Property
+noDoubleSpending = property $ do
+  t <- forAll (trace @UTXO 500)
+  let
+    utxo0 = fst . firstAndLastState $ t :: State UTXO
+    txs = traceSignals OldestFirst t :: [Signal UTXO]
+  when (all (noCommonInputs $ utxo utxo0) txs) $ do
+    (all (noCommonInputsTxs txs) (zip txs [0 .. ])) === True
```

Comment:
For the remaining checks, I think it's better to use `traverse_` and `===` so that when the condition fails we can get an error message telling us which one was the transaction that was duplicated (and we can look this up in the resulting trace). So removing the redundant `do` and `ins` binding this would become:

```haskell
    traverse_ (noCommonInputsTxs txs) (zip txs [0 .. ])
   where
    noCommonInputsTxs :: MonadTest m => [Tx] -> (Tx, Int) -> m ()
    noCommonInputsTxs txs (tx, i) =
      traverse_ (\txj -> txins' txj ∩ txins' tx === empty) (take i txs)

```

Does that make sense?

Reply:
I like this! Incorporated it in an upcoming commit.